### PR TITLE
Camcog pipeline joining the party

### PR DESCRIPTION
**Pull request by ColinBD**
To run locally use: _python data_transfer/main.py TFA Newcastle_. It's actually site agnostic so you only need to run it once and all sites are covered. The site input param needs to be valid but it doesn't matter if it's Newcastle or one of the other sites.

Note: when testing I normally comment out line 40 on tfa.py in dags to avoid files being uploaded to DMP.

I still have to implement an effective whitelist as around 14 files contain ThinkFAST data but no CANTAB data. I suspect a labelling issue is the cause which we can resolve.

### CODE REVIEW

File: data_transfer/devices/thinkfast.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+class ThinkFastRecording:
+    """
+    Stores most relevant metadata for readable lookup.
+    """
+
+    id: str
+    device_id: str
+    user_id: str
+    start: datetime
+    end: datetime
```
Comment (jawrainey) [inline data_transfer/devices/thinkfast.py:54]: This appears to not be used?

**Review by jawrainey — COMMENTED** [review]
Great work @ColinBD! I've not tested it locally yet, but given we'll deploy it to dev later we can do that instead. I've reviewed the PR and made comments on the code where parts might be clearer to be consistent with the other pipelines 🚀

File: data_transfer/db/main.py [hunk 2]
```diff
@@ -39,6 +39,7 @@ def update_record(record: Record) -> None:
 
 
 def all_hashes() -> List[str]:
+    print("looking in the DB to get all records")
```
Comment (jawrainey) [inline data_transfer/db/main.py:42]: We probably don't want to log this, so could be removed?

File: data_transfer/devices/thinkfast.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        remaining = len(participants)
+        log.debug(f"Number of our participants found in CamCog's database: {remaining}")
+
+        # loop through our list of participants retreiving their test data
+        for participant in participants:
+            raw_records = thinkfast_api.get_participants_records(participant.guid)
+            if len(raw_records) != 1:
+                log.debug(
+                    "Wow, raw_rec length is not 1, it is: " + str(len(raw_records))
+                )
```
[the remark this answers is not in the document]
Comment (ColinBD) [reply data_transfer/devices/thinkfast.py:129]: Good suggestion. I've added this in.

File: data_transfer/devices/thinkfast.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+                elif record.meta["tfa_type"] == "ThinkFAST":
+                    path = (
+                        record.download_folder()
+                        / f"{record.manufacturer_ref}-ThinkFAST.json"
+                    )
+                else:
+                    path = (
+                        record.download_folder()
+                        / f"{record.manufacturer_ref}-unknown.json"
+                    )
```
[the remark this answers is not in the document]
Comment (ColinBD) [reply data_transfer/devices/thinkfast.py:171]: A proper line saver that one!

File: data_transfer/devices/thinkfast.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+            # create a formatted record
+            # print("length of raw_records[0]: " + str(len(raw_records[0])))
+            # LOOP THROUGH RAW_RECORDS[0] making an entry for each
+            for record in raw_records[0]:
+                try:
+                    newRec = self.format_record(record, participant)
+                    if newRec:
+                        all_recs.append(newRec)
+                except Exception:
+                    log.debug("failed to create this record")
```
[the remark this answers is not in the document]
Comment (ColinBD) [reply data_transfer/devices/thinkfast.py:141]: No, raw_rec["itemGroups"][0]["items"][0]["measureCode"] returns something for both ThinkFAST and CANTAB. It only fails once per 100 records or so, so I need to investigate further. I will do this manually by printing raw_rec in such cases and examining the json to see what is going on.

File: data_transfer/devices/thinkfast.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+class ThinkFast:
+    def __init__(self, study_site: StudySite):
+        """
+        Use study_site name to build auth as there are multiple sites/credentials.
+        """
+        self.study_site = study_site
+        self.device_type = utils.DeviceType.TFA
+
+    def __unknown_records(self, records: list) -> Dict:
```
[the remark this answers is not in the document]
Comment (ColinBD) [reply data_transfer/devices/thinkfast.py:76]: Yes, agreed.

File: data_transfer/devices/thinkfast.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+                + participant.guid
+                + " has "
+                + str(len(raw_records[0]))
+                + " total TFA records. Writing "
+                + str(len(unknown_records))
+                + " new records to the DB"
+            )
+            # push the new records into the DB
+            for record in unknown_records.values():
+                create_record(record)
```
[the remark this answers is not in the document]
Comment (ColinBD) [reply data_transfer/devices/thinkfast.py:155]: This is amazingly concise. I would normally split this over two lines so I can't get confused thinking I've assigned something that has been popped away (which I know isn't the case as it doesn't happen in that order).

[the remark this answers is not in the document]
Comment (ColinBD) [reply data_transfer/devices/thinkfast.py:155]: It was _data = record.meta.pop("full_data")_ that works

File: data_transfer/lib/thinkfast.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+                        f"CORRECTED AN ERROR USING THE ODDITIES DICT: {known_incorrect_ids[ideaId]}"
+                    )
+                    newID = known_incorrect_ids[ideaId]
+                else:
+                    continue
+            participants.append(Participant(newID, rec["subjectIds"][0], rec["id"]))
+        # increment offset by the retreival limit
+        parameters["offset"] = str(int(parameters["offset"]) + 100)
+        # got all the data or do we need to make more API calls?
+        if int(parameters["offset"]) > int(response.json()["total"]):
```
[the remark this answers is not in the document]
Comment (ColinBD) [reply data_transfer/lib/thinkfast.py:110]: It wasn't happy about this and threw an error. I've added back in the 'int' conversions for now.

File: data_transfer/lib/thinkfast.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+                    f"INVALID ID FOUND WITHIN THE THINKFAST RECORDS. ID = {ideaId}"
+                )
+                # check dictionary of oddities and if we have a hit do the replacement
+                if ideaId in known_incorrect_ids:
+                    print(
+                        f"CORRECTED AN ERROR USING THE ODDITIES DICT: {known_incorrect_ids[ideaId]}"
+                    )
+                    newID = known_incorrect_ids[ideaId]
+                else:
+                    continue
```
[the remark this answers is not in the document]
Comment (ColinBD) [reply data_transfer/lib/thinkfast.py:105]: Done

File: data_transfer/lib/thinkfast.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+        try:
+            response = requests.get(
+                f"{config.thinkfast_api_url}/visit",
+                params=parameters,
+                headers=headers_dict,
+                auth=(config.thinkfast_username, config.thinkfast_password),
+            )
+            response.raise_for_status()
+            # store the response
+            results.append(response.json()["records"])
```
[the remark this answers is not in the document]
Comment (ColinBD) [reply data_transfer/lib/thinkfast.py:46]: done

File: data_transfer/lib/thinkfast.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ColinBD) [reply data_transfer/lib/thinkfast.py:105]: Hmm, this is cool and I could get used to it but I've always been a fan of the single line responsibility principle. Okay, I'll get with the times and make the change.

**Pull request merged**